torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -156,10 +153,55 @@ void PropagateShapeOnNodeByRunningIt(Node* node, const std::vector<TensorType*>&
 
   JIT_ASSERT(stack.size() == node->outputs().size());
   for (size_t i = 0; i < stack.size(); ++i) {
-    node->outputs()[i]->inferTypeFrom(stack[i].toTensor());
+    // some ops may have mixed tensor/primitive outputs
+    // for primitives, we don't need to change the type because it is already
+    // its most constrained form.
+    if(stack[i].isTensor())
+      node->outputs()[i]->inferTypeFrom(stack[i].toTensor());
   }
 }
 
+// is it ok to try to run the op
+// If an input is a constant, then we assume that the input is valid
+// and we can try to run it.
+// Otherwise:
+// Integral typed _inputs_ are often an indicator that we're indexing into
+// a tensor, so we should special-case these ops in the shape propagation.
+// Additionally, passing in a zero representative tensor into an integer
+// division op causes divide-by-zero errors
+// _Outputs_ must be tensors or primtives
+// We will call inferTypeFrom on the tensors, and ignore the primitives.
+// However, we allow primitive returns because we want to support mixed
+// primitive/tensor outputs.
+
+bool isValidArgumentForRunning(Value* v) {
+  // allow constants
+  if(toIValue(v))
+    return true;
+  if(TensorType* tt = v->type()->cast<TensorType>()) {
+    return !at::isIntegralType(tt->scalarType());
+  }
+  return v->type()->isSubtypeOf(*FloatType::get());
+}
+bool isValidReturnForRunning(Value* v) {
+  return v->type()->isSubtypeOf(*DynamicType::get()) ||
+      v->type()->isSubtypeOf(*NumberType::get());
```

Comment:
This is supposed to forbid lists?

Reply:
It is a whitelist, so as not to do something silly when we add a new type.